File: src/main/java/org/panteleyev/pwdmanager/NoteViewer.java
/*
 Copyright © 2017-2024 Petr Panteleyev <petr@panteleyev.org>
 SPDX-License-Identifier: BSD-2-Clause
 */
package org.panteleyev.pwdmanager;

import javafx.scene.control.ContextMenu;
import javafx.scene.control.TextArea;
import javafx.scene.input.Clipboard;
import javafx.scene.input.ClipboardContent;

import static org.panteleyev.fx.FxUtils.fxString;
import static org.panteleyev.fx.MenuFactory.menuItem;
import static org.panteleyev.pwdmanager.Constants.UI_BUNDLE;
import static org.panteleyev.pwdmanager.Shortcuts.SHORTCUT_C;
import static org.panteleyev.pwdmanager.bundles.Internationalization.I18N_COPY;

final class NoteViewer extends TextArea {
    public NoteViewer() {
        setEditable(false);
        setContextMenu(createContextMenu());
        setFocusTraversable(false);
    }

    private ContextMenu createContextMenu() {
        return new ContextMenu(
                menuItem(fxString(UI_BUNDLE, I18N_COPY), SHORTCUT_C, _ -> copy())
        );
    }

    @Override
    public void copy() {
        var text = getSelectedText();
        if (text.isEmpty()) {
            text = getText();
        }

        if (!text.isEmpty()) {
            var clipboard = Clipboard.getSystemClipboard();
            var ct = new ClipboardContent();
            ct.putString(text);
            clipboard.setContent(ct);
        }
    }
}

File: src/main/kotlin/org/panteleyev/pwdmanager/NoteViewer.kt
/*
 * Copyright (c) 2017, Petr Panteleyev <petr@panteleyev.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package org.panteleyev.pwdmanager

import javafx.scene.control.ContextMenu
import javafx.scene.control.MenuItem
import javafx.scene.control.TextArea
import javafx.scene.input.Clipboard
import javafx.scene.input.ClipboardContent
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyCodeCombination
import javafx.scene.input.KeyCombination

internal class NoteViewer : TextArea() {
    init {
        isEditable = false
        contextMenu = createContextMenu()
        isFocusTraversable = false
    }

    private fun createContextMenu(): ContextMenu {
        val menu = ContextMenu()

        val m1 = MenuItem("Copy...")
        m1.accelerator = KeyCodeCombination(KeyCode.C, KeyCombination.SHORTCUT_DOWN)
        m1.setOnAction { _ -> copy() }

        menu.items.addAll(m1)
        return menu
    }

    override fun copy() {
        var text = selectedText
        if (text.isEmpty()) {
            text = getText()
        }

        if (!text.isEmpty()) {
            val clipboard = Clipboard.getSystemClipboard()
            val ct = ClipboardContent()
            ct.putString(text)
            clipboard.setContent(ct)
        }
    }
}